modules/caddyhttp/reverseproxy/metrics.go

```diff
@@ -0,0 +1,80 @@
+package reverseproxy
+
+import (
+	"runtime/debug"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.uber.org/zap"
+)
+
+var reverseProxyMetrics = struct {
+	init             sync.Once
+	upstreamsHealthy *prometheus.GaugeVec
+}{
+	init: sync.Once{},
+}
+
+func initReverseProxyMetrics() {
+	const ns, sub = "caddy", "reverseproxy"
+
+	upstreamsLabels := []string{"upstream"}
+	reverseProxyMetrics.upstreamsHealthy = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "upstreams_healthy",
+		Help:      "Health status of reverse proxy upstreams.",
+	}, upstreamsLabels)
+}
+
+type metricsUpstreamsHealthyUpdater struct {
+	handler *Handler
+}
+
+func newMetricsUpstreamsHealthyUpdater(handler *Handler) *metricsUpstreamsHealthyUpdater {
+	reverseProxyMetrics.init.Do(func() {
+		initReverseProxyMetrics()
+	})
+
+	return &metricsUpstreamsHealthyUpdater{handler}
+}
+
+func (m *metricsUpstreamsHealthyUpdater) Init() {
```

Comment:
We do have one in the goroutine. Since the `Init` function only contains a single goroutine call, I don't see the point to add a recover() in the `Init` function itself. I might misunderstand something 😄

Reply:
Oh; wow, not sure how I missed that.